### PR TITLE
Additional improvements to Hathora SDK plugin

### PR DESCRIPTION
**Pull request by mikeseese**
This PR serves as an iterative improvement on top of #3 by @BMorinDrifter:
- Separates the API calls into multiple C++ files. While this is sometimes a controversial tactic, I believe the lower level ping logic in #3 detracts from from the readability by throwing everything in a single file. Further, this structure can keep everything under a single module for easy usage while keeping the plugin easier to maintain
- Adds a function for `GetPingServiceEndpoints` that only returns the endpoints, and allows developers to only retrieve pings for a subset of those endpoints. The PR also allows developers to provide an empty array to `GetRegionalPings` to retrieve pings for all regions to maintain the ease of use of the original implementation.
- Adds blueprint support for the current functions

### CODE REVIEW

**Comment by BMorinDrifter** [comment]
Overall, like the Blueprint library add.  
Unclear on the use case for not using Hathora's full ping set.  Perhaps something like a list of regions to ignore would be a better fit to intent?
Getting the libwebsocket C stuff into it's own file and away from the Unreal C++ stuff is a reasonable change, but the this SDK is very small + simple and not a fan of it being well on the road to dozens of very small files.  DiscoveredPingEndpoint.h would have been in a .cpp file it it wasn't for Unreal Header Tool.

**Comment by mikeseese** [comment]
> Unclear on the use case for not using Hathora's full ping set. Perhaps something like a list of regions to ignore would be a better fit to intent?

I don't believe that a developer-focused SDK module should be guard railing access to the underlying API; it leads to future feature requests for more generalization. I believe `GetPingServiceEndpoints` should only do the API call, and I also believe that `GetRegionalPings` shouldn't be calling the API call; I left it as a form of back-compatibility for those that don't want to be bothered with multiple callbacks. However, some use cases for the more flexible structure are:
- Having a whitelist of regions for some form of a regional rolling update (sure, I can agree that this is an odd scenario, but I don't think we should block it off because of an opinionated idea of what matchmaking looks like)
- Client could only do follow up pings to the closest N regions
- Having a whitelist that the client decides they're only interested in (i.e. I don't want to be put in matches outside of US West, so why should the client even try pinging the other endpoints?)
- Having a blacklist of regions to create fuller matches
- In my particular use case, the backend calls `GetPingServiceEndpoints` and decides what the clients should ping. The clients would only call `GetRegionalPings` with the array provided by backend
- Preventing a redundant HTTP request to the `GetPingServiceEndpoints` API endpoint for pinging several times (my client pings the regions every 10 seconds; while the extra request is minimal, it's unnecessary)

>  not a fan of it being well on the road to dozens of very small files.

I'm fine to agree to disagree. Since it is the Unreal way to favor larger, consolidated files, I have made the changes to keep all of the main module functions in HathoraSDK.cpp, kept the low level ping utility in `HathoraSDK_Ping[.h/.cpp]` files, and consolidated the USTRUCTs in `HathoraSDKTypes.h`.

**Comment by BMorinDrifter** [comment]
Second pass looks better after a quick glance.  Will pull it down later today and take a look since the diff is pretty hard to follow at this point.

>I don't believe that a developer-focused SDK module should be guard railing access to the underlying API; it leads to future feature requests for more generalization.

The Gamelift Server SDK doesn't surface the details how how it manages communication with the service. The broader AWS SDK has internal details that aren't surfaced or hidden (via optional parameters, builders, etc.) and this works out great.  It provides a clear contract and exposes that contract across multiple languages and a CLI with some minimal adjustments to fit particular languages and ecosystems.  Finally, these customer focused interfaces tend to be very stable even as the underlying service evolves.

As a customer what I'm looking for from this API is a set of community best practices around these APIs.  Things like retry policy, glue for all the components of ping collection, etc.  When/if Auth gets revisited, making that easy to migrate.  As a thin REST API wrapper there is a lot less value.

**Comment by mikeseese** [comment]
> As a customer what I'm looking for from this API is a set of community best practices around these APIs. Things like retry policy, glue for all the components of ping collection, etc. When/if Auth gets revisited, making that easy to migrate. As a thin REST API wrapper there is a lot less value.

I understand the value of having the set of community best practices available, but I don't believe that it should purposely omit an API that will force users that want it to implement the logic without things like retry logic or modify the plugin source to have to rebase when there are updates. It's a simple thing to add more options and keep the plugin unopinionated, while still allowing users to have the simple options. I don't have a heavy stake in the matter, and I'm happy to revert these changes just to add the blueprint library. I believe these changes would have been useful for me, and I'm sure others would too.

**Comment by BMorinDrifter** [comment]
My remaining concerns:

- HathoraSDKBPLibray.h puts the types FPingServiceEndpointsDynamicDelegate, FOnGetRegionalPingsDynamicDelegate and FConnectionInfoDynamicDelegate into the global namespace.  Consider either moving them into UHathoraSDKBPLibrary or giving putting Hathora in their name so they don't collide or get confused for something else.

- FHathoraSDKModule::GetRegionalPings and FHathoraSDKModule::GetRegionalPings_Internal have the same signature, is this intentional?  Suggest keeping the old signature for the prior function name and putting the additional functionality in the new function.

- It would be nice for the Blueprint Ping API to be a simple, single call with just a callback for the ping list like the original C++ API.  Blueprint authors are going to be less inclined to do heavy customization.

- HathoraSDK_Ping.h can be folded into HathoraSDK_Ping.cpp.  Nothing in there should be consumed by any other .cpp files

**Comment by mikeseese** [comment]
I implemented all of the feedback except regarding to the `FHathoraSDKModule::GetRegionalPings_Internal` function signature. 
 The Internal function is necessary to prevent an infinite loop in an odd scenario if Hathora's API ever returns an empty list of endpoints. I chose to have an Internal function as an alternative to having some sort of additional flag parameter. The matching signature is intentional for this reason. However, the Internal function should not have been `public` and has been moved to `protected`.

**Comment by mikeseese** [comment]
Closing as https://github.com/hathora/hathora-unreal-sdk/commit/25a27c171cc26cdb2b6e587ceb96045658521736 makes this PR obsolete
